components/dashboard/src/settings/Integrations.tsx

```diff
@@ -199,77 +179,435 @@ function GitProviders() {
+                            &nbsp;
+                        </div>
+                    </div>
+                    <div className="p-0 my-auto flex flex-col w-3/12">
+                        <span className="my-auto font-medium truncate overflow-ellipsis">{ap.type}</span>
                     </div>
-                    <div className="self-center">
-                        <ContextMenu menuEntries={gitProviderMenu(ap)}>
-                            <img className="w-8 h-8 p-1" src={ThreeDots} alt="Actions" />
-                        </ContextMenu>
+                    <div className="p-0 my-auto flex flex-col w-7/12">
+                        <span className="my-auto truncate text-gray-500 overflow-ellipsis">{ap.host}</span>
+                    </div>
+                    <div className="my-auto flex w-1/12">
+                        <div className="self-center hover:bg-gray-200 rounded-md cursor-pointer w-8">
+                            <ContextMenu menuEntries={gitProviderMenu(ap)}>
+                                <img className="w-8 h-8 p-1" src={ThreeDots} alt="Actions" />
+                            </ContextMenu>
+                        </div>
                     </div>
                 </div>
             ))}
         </div>
+        {providers && providers.length > 0 && (
+            <div className="flex justify-start mt-6">
+                <button className="h-full" onClick={() => setModal({ mode: "new" })}>New Git Integration</button>
+            </div>
+        )}
     </div>);
 }
 
+function GitIntegrationModal(props: ({
+    mode: "new",
+} | {
+    mode: "edit",
+    provider: AuthProviderEntry
+}) & {
+    userId: string,
+    onClose?: () => void
+    update?: () => void
+}) {
+
+    const callbackUrl = (host: string) => {
+        const pathname = `/auth/${host}/callback`;
+        return gitpodHostUrl.with({ pathname }).toString();
+    }
+
+    const [type, setType] = useState<string>("GitLab");
+    const [host, setHost] = useState<string>("gitlab.example.com");
+    const [redirectURL, setRedirectURL] = useState<string>(callbackUrl("gitlab.example.com"));
+    const [clientId, setClientId] = useState<string>("");
+    const [clientSecret, setClientSecret] = useState<string>("");
+    const [busy, setBusy] = useState<boolean>(false);
+    const [errorMessage, setErrorMessage] = useState<string | undefined>();
+    const [validationError, setValidationError] = useState<string | undefined>();
+
+    const firstRender = useRef(true);
+
+    useEffect(() => {
+        if (props.mode === "edit") {
+            setType(props.provider.type);
+            setHost(props.provider.host);
+            setClientId(props.provider.oauth.clientId);
+            setClientSecret(props.provider.oauth.clientSecret);
+            setRedirectURL(props.provider.oauth.callBackUrl);
+        }
+    }, []);
```

Comment:
As to our discussion yesterday the commonly used pattern here is this https://github.com/facebook/react/issues/15523#issuecomment-487820357

```
const [type, setType] = useState<string>(props.provider.type);
...

userEffect(() => {
   setType(props.provider.type);
   ...
}, [props.provider]);
```

Reply:
thanks for sharing! 
in this case, the `GitIntegrationModal` component can be used in two modes, with and without a `prop.provider` to allow creating and editing.